### PR TITLE
Use less stack space while parsing if-statements

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This gives us the ability to parse about 30 more frames during heavily recursive scenarios.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @RikkiGibson @jcouv @cston

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -7709,19 +7711,28 @@ private GotoStatementSyntax ParseGotoStatement()
...
-            var statement = firstTokenIsElse ? this.ParseExpressionStatement() : this.ParseEmbeddedStatement();
-            var elseClause = this.ParseElseClauseOpt();
+            return _syntaxFactory.IfStatement(
+                this.EatToken(SyntaxKind.IfKeyword),
+                this.EatToken(SyntaxKind.OpenParenToken),
+                this.ParseExpressionCore(),
+                this.EatToken(SyntaxKind.CloseParenToken),
+                this.ParseEmbeddedStatement(),
+                this.ParseElseClauseOpt());
+        }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: optimie the fast path of a normal if-statement, with no need for flow control or storage for flow control temps.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -7709,19 +7711,28 @@ private GotoStatementSyntax ParseGotoStatement()
...
+
+        private IfStatementSyntax ParseMisplacedElse()
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: MisplacedElse is a very rare case (usually just in the IDE as people are typing).  it's fine for it to have it's own dedicated parsing method.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -6444,8 +6444,10 @@ private StatementSyntax ParseStatementNoDeclaration(bool allowAnyExpression)
                 case SyntaxKind.GotoKeyword:
                     return this.ParseGotoStatement();
                 case SyntaxKind.IfKeyword:
-                case SyntaxKind.ElseKeyword: // Including 'else' keyword to handle 'else without if' error cases 
                     return this.ParseIfStatement();
+                case SyntaxKind.ElseKeyword:
+                    // Including 'else' keyword to handle 'else without if' error cases 
+                    return this.ParseMisplacedElse();
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: good idea!

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -7709,19 +7711,28 @@ private GotoStatementSyntax ParseGotoStatement()
...
+        }
+
+        private IfStatementSyntax ParseMisplacedElse()
+        {
+            Debug.Assert(this.CurrentToken.Kind == SyntaxKind.ElseKeyword);
 
-            return _syntaxFactory.IfStatement(@if, openParen, condition, closeParen, statement, elseClause);
+            return _syntaxFactory.IfStatement(
+                this.EatToken(SyntaxKind.IfKeyword, ErrorCode.ERR_ElseCannotStartStatement),
+                this.EatToken(SyntaxKind.OpenParenToken),
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Should we call EatToken, ParseExpressionCore, etc in this way even though we know they will always fail? Basically everything before 'ParseElseClauseOpt'?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Yes. That's the idea (and what the code did before). It effectively synthesizes the missing nodes/tokens we want here

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**